External/FEXCore/Source/Interface/Core/Core.cpp

```diff
@@ -1156,6 +1276,87 @@ namespace FEXCore::Context {
     }
   }
 
+  static void ClearMemoryInternal(std::map<uint64_t, Context::MemoryEntry> &MemoryMaps, uintptr_t Base, uintptr_t Size) {
+    const auto Top = Base + Size;
+
+    // find the first Mapping at or after the Range ends, or ::end()
+    // Top is the first value after the end mapping
+    auto Mapping = MemoryMaps.lower_bound(Top);
+
+    // Iterate backwards all mappings
+    while (Mapping != MemoryMaps.begin()) {
+      Mapping--;
+
+      const auto MapBase = Mapping->first;
+      const auto MapTop = MapBase + Mapping->second.Length;
+
+      if (MapTop <= Base) {
+        // Mapping ends before the Range start, exit
+        break;
+      } else if (MapBase < Base && MapTop <= Top) {
+        // Mapping starts before Range & ends at or before Range, trim end
+        Mapping->second.Length = Base - MapBase;
+      } else if (MapBase < Base && MapTop > Top) {
+        // Mapping starts before Range & ends after Range, split
+
+        // trim first half
+        Mapping->second.Length = Base - MapBase;
+
+        // insert second half
+        MemoryMaps.emplace(Top, Context::MemoryEntry{ MapTop - Top, Mapping->second.Writable });
+      } else if (MapBase >= Base && MapTop <= Top) {
+        // Mapping is included or equal to Range, delete
+        // returns next element, so -- is safe at loop
+        Mapping = MemoryMaps.erase(Mapping);
+      } else if (MapBase >= Base && MapTop > Top) {
+        // Mapping starts after or at Range && ends after Range, trim start
+
+        // insert second half
+        MemoryMaps.emplace(Top, Context::MemoryEntry{ MapTop - Top, Mapping->second.Writable });
+
+        // erase original
+        // returns next element, so it can be decremented safely in the next loop iteration
+        Mapping = MemoryMaps.erase(Mapping);
+      } else {
+        ERROR_AND_DIE_FMT("Invalid Case");
+      }
+    }
+  }
+
+  void Context::SetMemoryMap(uintptr_t Base, uintptr_t Size, bool Writable) {
+    Size = FEXCore::AlignUp(Size + (Base & ~FHU::FEX_PAGE_MASK), FHU::FEX_PAGE_SIZE);
+    Base = Base & FHU::FEX_PAGE_MASK;
+    
+    {
+      FHU::ScopedSignalMaskWithSharedMutex lk(MemoryEntryMutex, false);
+
+      // remove previous mappings
+      ClearMemoryInternal(MemoryMaps, Base, Size);
+
+      // Insert new mapping
+      MemoryMaps.emplace(Base, MemoryEntry{ Size, Writable });
+    }
+
+  }
+
+  void Context::ClearMemoryMap(uintptr_t Base, uintptr_t Size) {
+    Size = FEXCore::AlignUp(Size + (Base & ~FHU::FEX_PAGE_MASK), FHU::FEX_PAGE_SIZE);
+    Base = Base & FHU::FEX_PAGE_MASK;
+
+    {
+      FHU::ScopedSignalMaskWithSharedMutex lk(MemoryEntryMutex, false);
+      ClearMemoryInternal(MemoryMaps, Base, Size);
+    }
+  }
+
+  void Context::LockBeforeFork() {
+    MemoryEntryMutex.lock_shared();
```

Comment:
Shouldn't this do an exclusive lock? Otherwise, the problem you described about another thread holding the lock before `fork()`ing would still apply.

Reply:
Yes it should.